### PR TITLE
MODELIX-644 Support writing in generated typed API with ModelQl

### PR DESCRIPTION
**Pull request by mhuster23**
Adds the ability to perform write operations in ModelQL queries.
As the MetaModelGenerator class has organically grown to a significant size, [MODELIX-667](https://issues.modelix.org/issue/MODELIX-667) was created as a followup ticket.

### CODE REVIEW

File: modelql-typed/src/commonMain/kotlin/org/modelix/modelql/typed/TypedModelQL.kt [hunk 1]
```diff
@@ -140,6 +144,17 @@ object TypedModelQL {
...
+    fun <ParentT : ITypedNode, ChildT : ITypedNode> addNewChild(input: IMonoStep<ParentT>, link: ITypedChildListLink<ChildT>, index: Int = -1, concept: ITypedConcept): IMonoStep<ChildT> {
+        val conceptRef = ConceptReference(concept.untyped().getUID())
+        return input.untyped().addNewChild(link.untyped(), index, conceptRef).typedUnsafe(link.getTypedChildConcept().getInstanceInterface())
+    }
+
+    fun <ParentT : ITypedNode, ChildT : ITypedNode> setChild(input: IMonoStep<ParentT>, link: ITypedSingleChildLink<ChildT>, concept: ITypedConcept): IMonoStep<ChildT> {
+        val conceptRef = ConceptReference(concept.untyped().getUID())
+        input.untyped().children(link.untyped().key()).firstOrNull().mapIfNotNull { it.remove() }
+        return input.untyped().addNewChild(link.untyped(), conceptRef).typedUnsafe(link.getTypedChildConcept().getInstanceInterface())
+    }
```
Comment (languitar) [inline modelql-typed/src/commonMain/kotlin/org/modelix/modelql/typed/TypedModelQL.kt:156]: What about tests for these new methods?

File: model-api-gen/src/main/kotlin/org/modelix/metamodel/generator/MetaModelGenerator.kt [hunk 2]
```diff
@@ -463,6 +464,42 @@ class MetaModelGenerator(
...
                                     .build(),
                             )
+                            val returnType = IMonoStep::class.asTypeName().parameterizedBy(targetType)
+                            val receiverType = IMonoStep::class.asTypeName().parameterizedBy(concept.nodeWrapperInterfaceType())
+                            if (feature.multiple) {
+                                addFunction(
+                                    FunSpec.builder(feature.addMethodName())
+                                        .returns(returnType)
+                                        .receiver(receiverType)
+                                        .addParameter("concept", ITypedConcept::class.asTypeName())
```
Comment (languitar) [inline model-api-gen/src/main/kotlin/org/modelix/metamodel/generator/MetaModelGenerator.kt:474]: Can we use a default for this parameter when the child type is not abstract?

**Review by languitar — COMMENTED** [review]
Some initial comments inline. What about a test for the generated methods?

File: modelql-typed/src/commonMain/kotlin/org/modelix/modelql/typed/TypedModelQL.kt (hunk 1 above)
Comment (languitar) [reply modelql-typed/src/commonMain/kotlin/org/modelix/modelql/typed/TypedModelQL.kt:156]: Ah, separate commit.

File: model-api-gen/src/main/kotlin/org/modelix/metamodel/generator/GeneratorInput.kt [hunk 3]
```diff
@@ -302,7 +302,10 @@ internal sealed class ProcessedLink(name: String, uid: String?, optional: Boolea
 }
 
 internal class ProcessedChildLink(name: String, uid: String?, optional: Boolean, var multiple: Boolean, type: ProcessedConceptReference, deprecationMessage: String?) :
-    ProcessedLink(name, uid, optional, type, deprecationMessage)
+    ProcessedLink(name, uid, optional, type, deprecationMessage) {
+
+    fun addMethodName() = "add" + generatedName.take(1).uppercase() + generatedName.drop(1)
```
Comment (slisson) [inline model-api-gen/src/main/kotlin/org/modelix/metamodel/generator/GeneratorInput.kt:307]: Maybe something like `adderMethodName` would be less confusing.

File: model-api-gen-gradle-test/kotlin-generation/src/test/kotlin/org/modelix/modelql/typed/TypedModelQLTest.kt [hunk 4]
```diff
@@ -195,4 +236,167 @@ class TypedModelQLTest {
...
+        assertEquals(C_StaticMethodDeclaration.untyped().getUID(), children.last().untyped().concept?.getUID())
+    }
+
+    @Test
+    fun `add new child - default concept`() = runTest { httpClient ->
+        val client = ModelQLClient.builder().url("http://localhost/query").httpClient(httpClient).build()
+        client.query { root ->
+            root.descendants().ofConcept(C_XmlComment)
+                .first()
+                .addLines()
```
Comment (slisson) [inline model-api-gen-gradle-test/kotlin-generation/src/test/kotlin/org/modelix/modelql/typed/TypedModelQLTest.kt:355]: The generated method name is a bit confusing in cases like this. Maybe `addTo` would be a better prefix?

**Comment by mhuster23** [comment]
> Do we have automatic disambiguation in case an adder method name would collide with a different accessor name from the model?

No, but I think we can assume that there is no collision. It would be a weird model, if a role `x` and `addToX` existed both on the same concept at the same time.

**Comment by slisson** [comment]
Conflicts between role names are already resolved here: https://github.com/modelix/modelix.core/blob/9f267dfbcb2692cdd9967319c7aab84ea8399284/model-api-gen/src/main/kotlin/org/modelix/metamodel/generator/GeneratorInput.kt#L94-L137
`generatedName` is guaranteed to be unique.

**Comment by slisson** [comment]
I think it's save to assume that there will be no role starting with `addTo`.

**Review by slisson — APPROVED** [review]

**Pull request merged**